lib/lists.nix

```diff
@@ -612,6 +612,40 @@ rec {
     # Input list
     list: sublist count (length list) list;
 
+  /* Whether the first list is a prefix of the second list.
+
+  Type: hasPrefix :: [a] -> [a] -> bool
+
+  Example:
+    hasPrefix [ 1 2 ] [ 1 2 3 4 ]
+    => true
+    hasPrefix [ 0 1 ] [ 1 2 3 4 ]
+    => false
+  */
+  hasPrefix =
+    list1:
+    list2:
+    take (length list1) list2 == list1;
+
+  /* Remove the first list as a prefix from the second list.
+  Error if the first list isn't a prefix of the second list.
+
+  Type: removePrefix :: [a] -> [a] -> [a]
+
+  Example:
+    removePrefix [ 1 2 ] [ 1 2 3 4 ]
+    => [ 3 4 ]
+    removePrefix [ 0 1 ] [ 1 2 3 4 ]
+    => <error>
+  */
+  removePrefix =
+    list1:
+    list2:
+    if hasPrefix list1 list2 then
+      drop (length list1) list2
+    else
+      throw "lib.lists.removePrefix: First argument is not a list prefix of the second argument";
```

Comment:
this behavior could be surprising if the user is already familiar with `lib.strings.removePrefix`

Reply:
Oh that's a good point, though I'd argue that this here is the more correct behavior, at least given the name of this function. A version that doesn't fail should maybe be called `tryRemovePrefix` instead.

Reply:
```haskell
ghci> :m + Data.List
ghci> stripPrefix "hi" "bar"
Nothing
```

Was hoping for resolution, but `null` would be the third option.

Reply:
Or avoid completely and use continuation passing.

`handlePrefix :: { prefix :: [a], onSuffix :: [a] -> r, onMismatch :: r } -> [a] -> r`

Reply:
Continuations would be a bit unusual for `lib`. I like the `null` option more, Rust also [does the same](https://doc.rust-lang.org/std/primitive.slice.html#method.strip_prefix)

Reply:
And then arguably `hasPrefix` is barely necessary anymore, because you can just do `removePrefix list != null`, which would then also be more efficient.

However, the path version of these functions already exist now with the `throw` semantics: https://github.com/NixOS/nixpkgs/pull/237610 and https://github.com/NixOS/nixpkgs/pull/238013